tutorials/Numerics/LinearSolvers/bgmres.jl

```diff
@@ -0,0 +1,191 @@
+            for i4 in 1:tup[4]
+                for i2 in 1:tup[2]
+                    for i1 in 1:tup[1]
+                        tmp = alias_x[i1, i2, :, i4, :, i6][:]
+                        tmp2 = A[i1, i2, i4, i6] * tmp
+                        alias_y[i1, i2, :, i4, :, i6] .=
+                            reshape(tmp2, (tup[3], tup[5]))
+                    end
+                end
+            end
+        end
+    end
+end
+# Next we define the array structure of an MPIStateArray
+# in its true high dimensional form
+tup = (2, 2, 5, 2, 10, 2);
+# We define our linear operator as a random matrix
+Random.seed!(1234)
+B = [
+    randn(tup[3] * tup[5], tup[3] * tup[5])
+    for i1 in 1:tup[1], i2 in 1:tup[2], i4 in 1:tup[4], i6 in 1:tup[6]
+];
+columnwise_A = [
+    B[i1, i2, i4, i6] + 3*(i1+i2+i4+i6)*I
+    for i1 in 1:tup[1], i2 in 1:tup[2], i4 in 1:tup[4], i6 in 1:tup[6]
+];
+# as well as its inverse
+columnwise_inv_A = [
+    inv(columnwise_A[i1, i2, i4, i6])
+    for i1 in 1:tup[1], i2 in 1:tup[2], i4 in 1:tup[4], i6 in 1:tup[6]
+];
+columnwise_linear_operator! = closure_linear_operator!(columnwise_A, tup);
+columnwise_inverse_linear_operator! =
+    closure_linear_operator!(columnwise_inv_A, tup);
+# The structure of an MPIStateArray is related to its true
+# higher dimenionsal form as follows:
+mpi_tup = (tup[1] * tup[2] * tup[3], tup[4], tup[5] * tup[6])
+# We now define the right hand side of our Linear system
+b = randn(mpi_tup)
+# As well as the initial guess
+x = copy(b)
+x += randn(mpi_tup) * 0.1
+# In the previous tutorial we mentioned that it is assumed that
+# the right hand side is an array whose column vectors all independent linear
+# systems. But right now the array structure of ``x`` and ``b`` do not follow
+# this requirement.
+# To handle this case we must pass in additional arguments that tell the
+# linear solver how to reconcile these differences.
+# The first thing that the linear solver must know of is the higher tensor
+# form of the MPIStateArray, which is just the `tup` from before
+reshape_tuple_f = tup
+# The second thing it needs to know is which indices correspond to a column
+# and we want to make sure that these are the first set of indices that appear
+# in the permutatation tuple (which can be thought of as enacting
+# a Tensor Transpose).
+permute_tuple_f = (5, 3, 4, 6, 1, 2) # make the column indices the fast indices
+# It has this format since the 3 and 5 index slots
+# are the ones associated with traversing a column. And the 4 index
+# slot corresponds to a state.
+# We also need to tell our solver which kind of Array struct to use
+ArrayType = Array
+
+# We are now ready to finally define our linear solver, which uses a number
+# of keyword arguments
+gmres = BatchedGeneralizedMinimalResidual(b, ArrayType = ArrayType, m = tup[3]*tup[5]*tup[4], n = tup[1]*tup[2]*tup[6], reshape_tuple_f = reshape_tuple_f, permute_tuple_f = permute_tuple_f, atol = eps(Float64)*10^2, rtol = eps(Float64)*10^2)
```

Comment:
Can we flatten this line out?

Reply:
I think the formatter should catch it when it is run